core/base/validation.hpp

```diff
@@ -0,0 +1,183 @@
+}
+
+
+template <typename ValueType>
+bool is_finite(const gko::array<ValueType>& values)
+{
+    const auto host_values =
+        make_temporary_clone(values.get_executor()->get_master(), &values);
+
+    for (size_t i = 0; i < host_values->get_size(); ++i) {
+        if (!is_finite_scalar(host_values->get_const_data()[i])) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+
+template <typename IndexType>
+bool has_unique_ptrs(const gko::array<IndexType>& row_ptrs,
+                     const gko::array<IndexType>& col_idxs)
+{
+    const auto host_row_ptrs =
+        make_temporary_clone(row_ptrs.get_executor()->get_master(), &row_ptrs);
+    const auto host_col_idxs =
+        make_temporary_clone(col_idxs.get_executor()->get_master(), &col_idxs);
+
+    const auto num_rows_ = host_row_ptrs->get_size() - 1;
+    const auto row_ptrs_ = host_row_ptrs->get_const_data();
+    const auto col_idxs_ = host_col_idxs->get_const_data();
+
+    bool result = true;
+
+    for (IndexType row = 0; row < num_rows_; row++) {
+        const auto begin = row_ptrs_[row];
+        const auto end = row_ptrs_[row + 1];
+        const auto size = end - begin;
+        std::unordered_set<IndexType> unique_ptrs(col_idxs_ + begin,
+                                                  col_idxs_ + end);
+
+        result = result && unique_ptrs.size() == size;
+    }
+    return result;
+}
+
+template <typename IndexType>
+bool has_unique_perm_idxs(const gko::array<IndexType>& permutation_)
+{
+    const auto host_perm_idxs = make_temporary_clone(
+        permutation_.get_executor()->get_master(), &permutation_);
+
+    const auto size = host_perm_idxs->get_size();
+    std::unordered_set<IndexType> unique_ptrs(
+        host_perm_idxs->get_const_data(),
+        host_perm_idxs->get_const_data() + size);
+
+    return unique_ptrs.size() == size;
+}
+
+// template <typename ValueType, typename IndexType>
+// bool has_non_zero_diagonal(const gko::array<IndexType>& row_ptrs,
+//                            const gko::array<IndexType> col_idxs,
+//                            const gko::array<ValueType>& values)
+// {
+//     const auto host_row_ptrs =
+//         make_temporary_clone(row_ptrs.get_executor()->get_master(),
+//         &row_ptrs);
+//     const auto host_col_idxs =
+//         make_temporary_clone(col_idxs.get_executor()->get_master(),
+//         &col_idxs);
+//     const auto host_values =
+//         make_temporary_clone(values.get_executor()->get_master(), &values);
+
+//     const auto num_rows = host_row_ptrs->get_size() - 1;
+//     const auto row_ptrs = host_row_ptrs->get_const_data();
+//     const auto col_idxs = host_col_idxs->get_const_data();
+//     const auto values = host_values->get_const_data();
+
+//     if(row_ptrs->get_size() != col_idxs->get_size())
+//     {
+//         return false;
+//     }
+
+//     if (values->get_size() > num_rows * (num_rows -1))  //pigeonhole
```

Comment:
I think until we need them, we can remove these functions instead of leaving them commented out